### PR TITLE
Feature: Framerate display window

### PR DESCRIPTION
**Pull request by nielsmh**
Frame rate and various game loop/graphics timing measurements and graphs. Accessible via the Help menu, and can print some stats in the console via the `fps` command.

image

There should be hooks into all video output drivers, but platforms should be tested that the video output measurement is actually meaningful, and that the timer used for measurements is sufficiently accurate.

Tested working on:
- [x] Windows (Win32)
- [ ] Windows (SDL)
- [x] Mac
- [x] Linux (SDL)
- [ ] DOS (Allegro)
- [ ] OS/2

-----

Original first image:
image

### CODE REVIEW

**Comment by nielsmh** [comment]
Inspiration for this was reading #6677, having a framerate display that also measures individual components of game speed should make it easier to figure out "why is it not fast".

**Comment by stavrossk** [comment]
@nielsmh This is awesome, thank you!!!

**Comment by nielsmh** [comment]
Time for bikeshedding: Is this preferable to the style shown in the first screenshot?

image

**Comment by andythenorth** [comment]
As far as I can tell, compiles and works as expected on Mac.  I can't verify timing accuracy :)

**Comment by PeterN** [comment]
- Game loop shows frames/s even when paused.
- Video output shows red when running at game loop rate, and is affected hugely by mouse cursor movement.
- In fast-forward mode, frames/s switches between 500, 1000 and 9999.99 -- granularity issue?

**Comment by nielsmh** [comment]
1) I'll have to test more with pause mode, I haven't actually checked what happens then.
2) My assumption has been that you want video output to run at 60 fps, simply because of mouse cursor/UI interaction being better then. Mouse movement affects window update rate on Windows since it causes messages to enter the message loop more frequently than it would otherwise process screen updates.
3) On Windows, the timing is obtained with GetTickCount which doesn't have good granularity, so multiple calls within a 20 ms window (or there about) will often return the same value. Possibly a multimedia timer timestep increase and measurement via `timeGetTime` instead could give better accuracy.

**Comment by PeterN** [comment]
You appear to be timing how long a frame takes and storing lots of these values, to then calculate an average based on a number of date points, instead of time. This means the average takes quite a while to settle on a value.

You should probably just count how many frames happen within 1 second. This will reduce timing errors, require less CPU effort to calculate, and means the average value is always just from the last second.

**Comment by PeterN** [comment]
Hmm, of course, it is probably necessary to do it your way to be able to break it down into sections.

If (on display) you pick the number of data points that correspond to about 1 second that will fix the settle-time issue. Limiting the refresh rate of the fps window will reduce fluctuation too.
I guess you need to use high-resolution timers to get further accuracy.

**Comment by nielsmh** [comment]
An alternative approach could be to bucketize time, e.g. have 11 buckets of 100 ms intervals, and count number of frames in each interval in a bucket. So you have `bucket = (ms_timestamp / 100) % 11`, have one bucket that's "active", and calculate FPS from the sum of the remaining 10. Whenever the timestamp rolls over to the next 100 ms interval, the new bucket for that interval is then reset to 0 frames counted.

**Comment by TrueBrain** [comment]
I would really like such a feature. And I like the first window more than the second; but that is more because the second confuses me.

Lately in a lot of games you see two values: FPS and UPS. FPS being how often your screen is redrawn; so how many times changes appear on your screen in a single second. UPS being how often the game updates it states (updates per second). So our gameloop basically.

In modern games they almost always run in two different threads, making measuring a lot easier ofc :)

What I do not really get in your second screenshot, is video output. It feels off that you have both a screen redrawing and video output. Why are they different? Seems like a waste of CPU/GPU power if one is on a higher frequency than the other.

What would be really cool to add (possibly in another PR), is a history. Only knowing the current FPS means you constantly have to look at the number. Having a history gives you more time to react on jumps etc.

Lastly, what would be really really cool, is if in for example the UPS (GameLoop) you can see how much ms each step takes (profiling on a high-level). So you see how many ms vehicle-loop took, etc. Same for FPS. But this might not be for this PR :D

Measuring you see in a wide variety. You have those that check every second how many frames, and that number bounces up and down like a mofo. I never liked those values, as they don't really mean anything. That only works if you show them over time. This is the addition you often see on this. So instead of showing the FPS of that second, you also see the last 30 seconds or so. This allows you to see spikes, etc. Still, per second is a very unreliable way of looking at those numbers. (very spiky)

The other way around is also possible, where you check every N frames how many seconds it took. And than again with the history, etc. This is what you see most in simple games, as the implementation is cheapest. The downside is if your framerate drops to 0, it takes for ever for the FPS to update. But the update to see the FPS needs a new frame, so you might wonder if it matters :)

The bucket system you suggest is also a very solid one. Calculating which bucket is cheap, and it gives a sliding window (at least, I assumed you meant that). The upside is that the FPS is a lot more stable (fluctuations take a while to show up). Downside is that putting the resulting FPS in a history is non-sensible. But this can be Incorporated with the first suggestion for history, if you make the buckets for, say, 30 seconds (with a resolution of 100ms sounds fine). For the active FPS you can take a sliding window of 10 buckets or so, and show with 100ms resolution the measured FPS in a history. This only is vastly more difficult to implement correctly, so I often see projects just pick the second approach (per N frames) and be done with it.

Sorry for the long reply of which most you most likely already knew. But sometimes it is good to write it down :D

tldr; keep it simple, UPS/FPS, history if possible would be awesome :D

**Comment by nielsmh** [comment]
@TrueBrain The way I've implemented things it could easily set up to also measure more specific parts of the game loop. Just add another element to `enum FramerateElement` and instantiate a `FramerateMeasurer` object in the scope you want to measure timing of.

The reason I have video output as a separate element is that, on some platforms, the action of copying the composited frame into video memory might actually become the bottleneck, depending on resolution, bit depth, and maybe color-space conversion, and I want to be able to display that. The point of the video output element isn't so much the FPS as it is the average time taken in ms.

The framerate window could get an expand/collapse button (similar to the Finances window) so it shows just the game loop and screen drawing FPS by default, but can expand to a list of detailed timings of several elements of the game loop.

**Comment by nielsmh** [comment]
I just tried out making a timestamp-bucketized framerate measurement, and don't like it very much. It gives less precision unless I increase the number of buckets to something like 100, and then it won't have a good response time anyway. I think a better approach is to simply keep track of how many data points are valid and never use more than the valid count for calculating current rate.

[This gist has the failed bucketizer.](https://gist.github.com/nielsmh/2a800f795261bedc589995385ec13964)

**Comment by nielsmh** [comment]
I added some basic time graphs now.

image

Right now they're just using a fixed scaling (1 pixel in height = 1 millisecond measured) and lack any labels. The height is fixed to 100 pixels, with the assumption that any values above that are meaninglessly large. The X-axis timescale is simply one pixel per data point, i.e. per tick or frame.

**Comment by nielsmh** [comment]
Much improved graphs now, including support for scaling based on font size.

**Comment by TrueBrain** [comment]
I really really like what you did here @nielsmh . Well done! (did not look at the code, sorry; just the end-result :P)

**Comment by TrueBrain** [comment]
PS: we are going to get so much questions why it "only" runs at 32.25 frames/s :P

**Comment by nielsmh** [comment]
I really have no idea why it gets the framerate wrong, but I'm going to investigate it in detail today. Or perhaps the game really _is_ running slightly slow due to system timers or whatever?

**Comment by PeterN** [comment]
Not impossible as the system does not try to catch up, but unlikely as it is showing exactly 32.25 for everyone who has tested.

**Comment by nielsmh** [comment]
At least reading win32_v.cpp `MainLoop`, it looks like timing is based on GetTickCount, and whether to advance to next frame is determined by current timestamp >= previous frame timestamp + 30 ms. Since GetTickCount results are normally quantized to 16 ms or there about, I think there will be a bit of drift there, causing the actual frame rate to be somewhat lower than the intended 33 1/3.

**Comment by nielsmh** [comment]
I think it will be worth trying to use `timeGetTime()` for timing instead, paired with `timeBeginPeriod(MILLISECONDS_PER_TICK)`, to see if the framerate display gets closer to the intended.

**Comment by nielsmh** [comment]
Well you know what? The game really _was_ running at 32.25 fps, because of quantisation in GetTickCount. Changing it to use timeGetTime and a timer precision of MILLISECONDS_PER_TICK, the fps display actually locks at 33.33 as it ought to.

**Comment by nielsmh** [comment]
From discussions on IRC, I think this is just about feature complete now? Platform testing would be good.

**Comment by nielsmh** [comment]
Here we go, now "accumulating" measurements are also supported, so the time spent on individual vehicle types can be measured.

image

**Comment by nielsmh** [comment]
I guess this one is controversial: If the game is paused, should it keep recording zero-measurements and draw a zeroed graph the entire time, or should it put the measurements on pause as well and continue afterwards as if no time had passed, with no gaps in the graph? It means the game loop-related and non-game loop bound graphs get out of sync, if the "stasis" method is used.

File: src/framerate_gui.cpp [hunk 1]
```diff
@@ -0,0 +1,752 @@
...
+				} else {
+					count--;
+				}
+			}
+
+			if (count > 0) {
+				return sumtime * 1000 / count / TIMESTAMP_PRECISION;
+			} else {
+				return 0;
+			}
```
Comment (PeterN) [inline src/framerate_gui.cpp]: Can you duff this into:

if (count == 0) return 0;
return sumtime * 1000 / count / TIMESTAMP_PRECISION;

File: src/framerate_gui.cpp [hunk 2]
```diff
@@ -0,0 +1,752 @@
...
+				point--;
+				count++;
+				if (point < 0) point = NUM_FRAMERATE_POINTS - 1;
+			}
+
+			if (total > 0 && count > 0) {
+				return (double)count * TIMESTAMP_PRECISION / total;
+			} else {
+				return 0;
+			}
```
Comment (PeterN) [inline src/framerate_gui.cpp]: As above

File: src/framerate_gui.cpp [hunk 3]
```diff
@@ -0,0 +1,752 @@
...
+
+}
+
+
+static std::chrono::high_resolution_clock _hr_clock;
+static TimingMeasurement GetPerformanceTimer()
+{
+	auto usec = std::chrono::time_point_cast<std::chrono::microseconds>(_hr_clock.now());
+	return usec.time_since_epoch().count();
+}
```
Comment (PeterN) [inline src/framerate_gui.cpp]: _hr_clock is only used once. Can you not just call std::chrono::high_resolution_clock.now() directly?

File: src/framerate_gui.cpp [hunk 4]
```diff
@@ -0,0 +1,752 @@
...
+		this->InitNested(number);
+		this->small = this->IsShaded();
+	}
+
+	virtual void OnTick()
+	{
+		if (this->small != this->IsShaded()) {
+			this->small = this->IsShaded();
+			auto caption = this->GetWidget<NWidgetLeaf>(WID_FRW_CAPTION);
+			if (this->small) {
```
Comment (PeterN) [inline src/framerate_gui.cpp]: Maybe just do this->small ? STR_FRAMERATE_CAPTION_SMALL : STR_FRAMERATE_CAPTION

File: src/framerate_gui.cpp [hunk 5]
```diff
@@ -0,0 +1,752 @@
...
+	}
+
+	typedef double (ElementValueExtractor)(PerformanceElement elem);
+	void DrawElementTimesColumn(const Rect &r, int heading_str, ElementValueExtractor get_value_func) const
+	{
+		int y = r.top;
+		DrawString(r.left, r.right, y, heading_str, TC_FROMSTRING, SA_CENTER);
+		y += FONT_HEIGHT_NORMAL + VSPACING;
+
+		for (auto e = PFE_FIRST; e < PFE_MAX; e = (PerformanceElement)(e + 1)) {
```
Comment (PeterN) [inline src/framerate_gui.cpp]: Cast is odd, e++ should suffice?

**Comment by PeterN** [comment]
Also, any chance of toning down the use of auto a bit?

File: src/video/win32_v.cpp [hunk 6]
```diff
@@ -1182,7 +1190,7 @@ static void CheckPaletteAnim()
 void VideoDriver_Win32::MainLoop()
 {
 	MSG mesg;
-	uint32 cur_ticks = GetTickCount();
+	uint32 cur_ticks = timeGetTime();
```
Comment (PeterN) [inline src/video/win32_v.cpp]: These changes don't belong in this PR

File: src/framerate_gui.cpp (hunk 3 above)
Comment (nielsmh) [reply src/framerate_gui.cpp]: Some implementations might have state kept around in the clock object, in fact I had hoped Microsoft's had had that. (It calls QueryPerformanceFrequency every call to `now()` even though the frequency should never change while the system is running.) Of course it could also be a static variable inside the function.

File: src/framerate_gui.cpp (hunk 5 above)
Comment (nielsmh) [reply src/framerate_gui.cpp]: MSVC 2015 complains about ++ operator not being valid for the type.

> error C2678: binary '++': no operator found which takes a left-hand operand of type 'PerformanceElement' (or there is no acceptable conversion)

I do see that some other enum types have had a custom `operator++` implemented for them.

Comment (LordAro) [reply src/framerate_gui.cpp]: https://github.com/OpenTTD/OpenTTD/blob/master/src/core/enum_type.hpp#L16 ?

File: src/video/win32_v.cpp (hunk 6 above)
Comment (nielsmh) [reply src/video/win32_v.cpp]: Agree, it's not a complete solution either.

File: src/framerate_gui.cpp (hunk 3 above)
Comment (nielsmh) [reply src/framerate_gui.cpp]: Wait, didn't realize now() was a static function... huh.

**Comment by nielsmh** [comment]
Slightly odd effect with the viewport rendering times graph "spewing the values up", this is due to the graph rendering reading incomplete/zero values for the first item in graph, since viewport rendering happens in the same cycle as the graph drawing. Fixing this would be annoying.

**Comment by LordAro** [comment]
...how annoying? :)

**Comment by nielsmh** [comment]
Two parts to fixing it: First is that you'd need to record the start _and_ end of the cycle that measurements are accumulated over. Second is that the accumulation would have to either happen in another scratch area, or consume a slot from the back end of the durations buffer which is uncounted during accumulation.

File: src/framerate_type.h [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+enum PerformanceElement {
+	PFE_GAMELOOP,        ///< Speed of gameloop processing.
+	PFE_FIRST = PFE_GAMELOOP,
+	PFE_ACC_GL_ECONOMY,  ///< Time spent processing cargo movement
+	PFE_ACC_GL_TRAINS,   ///< Time spent processing trains
+	PFE_ACC_GL_ROADVEHS, ///< Time spend processing road vehicles
+	PFE_ACC_GL_SHIPS,    ///< Time spent processing ships
+	PFE_ACC_GL_AIRCRAFT, ///< Time spent processing aircraft
+	PFE_GL_LANDSCAPE,    ///< Time spent processing other world features
```
Comment (nielsmh) [inline src/framerate_type.h]: The Landscape element actually accumulates across three different functions, but not over a number of similar objects being processed. Should it have the ACC infix? Should the ACC infix even exist?

File: src/landscape.cpp [hunk 8]
```diff
@@ -1304,10 +1307,14 @@ void OnTick_LinkGraph();
...
+	{
+		PerformanceAccumulator framerate(PFE_GL_LANDSCAPE);
+
+		OnTick_Town();
+		OnTick_Trees();
+		OnTick_Station();
+		OnTick_Industry();
+	}
 
 	OnTick_Companies();
```
Comment (nielsmh) [inline src/landscape.cpp:1319]: I'm not sure if OnTick_Companies does anything interesting worth measuring.

File: src/lang/english.txt [hunk 9]
```diff
@@ -2696,6 +2697,57 @@ STR_ABOUT_ORIGINAL_COPYRIGHT                                    :{BLACK}Original
...
+STR_FRAMERATE_RATE_GAMELOOP                                     :{WHITE}Simulation rate: {STRING2}
+STR_FRAMERATE_RATE_GAMELOOP_TOOLTIP                             :{BLACK}Number of game ticks simulated per second.
+STR_FRAMERATE_RATE_BLITTER                                      :{WHITE}Graphics frame rate: {STRING2}
+STR_FRAMERATE_RATE_BLITTER_TOOLTIP                              :{BLACK}Number of video frames rendered per second.
+STR_FRAMERATE_SPEED_FACTOR                                      :{WHITE}Current game speed factor: {DECIMAL}x
+STR_FRAMERATE_SPEED_FACTOR_TOOLTIP                              :{BLACK}How fast the game is currently running, compared to the expected speed at normal simulation rate.
+STR_FRAMERATE_CURRENT                                           :{WHITE}Current
+STR_FRAMERATE_AVERAGE                                           :{WHITE}Average
+STR_FRAMERATE_DATA_POINTS                                       :{WHITE}Data based on {COMMA} measurements
+STR_FRAMERATE_VALUE                                             :{STRING2}
```
Comment (nielsmh) [inline src/lang/english.txt]: This feels bad, it's just an artifact of how the code is implemented and I think it would be better to clean up the code and take out this indirection?

File: src/framerate_gui.cpp [hunk 10]
```diff
@@ -0,0 +1,813 @@
...
+			EndContainer(),
+			NWidget(WWT_TEXT, COLOUR_GREY, WID_FRW_INFO_DATA_POINTS), SetDataTip(STR_FRAMERATE_DATA_POINTS, 0x0),
+		EndContainer(),
+	EndContainer(),
+};
+
+struct FramerateWindow : Window {
+	bool small;
+
+	static const int VSPACING = 3; ///< space between column heading and values
```
Comment (nielsmh) [inline src/framerate_gui.cpp]: This is a really bad name isn't it? Is there any standard constant usable instead?

File: src/framerate_gui.cpp [hunk 11]
```diff
@@ -0,0 +1,813 @@
...
+		}
+
+		this->SelectVerticalScale(peak_value);
+	}
+
+	virtual void OnTick()
+	{
+		this->SetDirty();
+
+		if (this->scale_update_timer > 0) {
```
Comment (nielsmh) [inline src/framerate_gui.cpp]: Considering basing this off real time instead of ticks, I think that might give better response times when performance is bad, and worse when performance is good.

File: src/framerate_gui.cpp [hunk 12]
```diff
@@ -0,0 +1,813 @@
...
+			this->scale_update_timer--;
+		} else {
+			this->scale_update_timer = 10;
+			this->UpdateScale();
+		}
+	}
+
+	/** Scale and interpolate a value from a source range into a destination range */
+	template<typename T>
+	static inline T Scinterlate(T dst_min, T dst_max, T src_min, T src_max, T value)
```
Comment (nielsmh) [inline src/framerate_gui.cpp:645]: Does this operation have a more standard name?

File: src/framerate_gui.cpp [hunk 13]
```diff
@@ -0,0 +1,813 @@
...
+		/** Duration value indicating the value is not valid should be considered a gap in measurements */
+		static const TimingMeasurement INVALID_DURATION = UINT64_MAX;
+
+		/** Time spent processing each cycle of the performance element, circular buffer */
+		TimingMeasurement durations[NUM_FRAMERATE_POINTS];
+		/** Start time of each cycle of the performance element, circular buffer */
+		TimingMeasurement timestamps[NUM_FRAMERATE_POINTS];
+		/** Expected number of cycles per second when the system is running without slowdowns */
+		double expected_rate;
+		/** Next index to write to in \c durations and \c timestamps */
```
Comment (nielsmh) [inline src/framerate_gui.cpp:40]: Is this usage of formatting codes correct?

File: src/framerate_gui.cpp [hunk 14]
```diff
@@ -0,0 +1,813 @@
...
+			/* Position of last rendered data point */
+			Point lastpoint = {
+				x_max,
+				(int)Scinterlate<int64>(y_zero, y_max, 0, this->vertical_scale, durations[point])
+			};
+			/* Timestamp of last rendered data point */
+			TimingMeasurement lastts = timestamps[point];
+
+			TimingMeasurement peak_value = 0;
+			Point peak_point{ 0, 0 };
```
Comment (nielsmh) [inline src/framerate_gui.cpp]: C++11 initialization, should be fixed.

File: src/console_cmds.cpp [hunk 15]
```diff
@@ -1894,6 +1894,18 @@ static void IConsoleDebugLibRegister()
 	IConsoleAliasRegister("dbg_echo2",      "echo %!");
 }
 #endif
+DEF_CONSOLE_CMD(ConFramerate)
+{
+	extern void ConPrintFramerate(); // framerate_gui.cpp
+
+	if (argc == 0) {
```
Comment (nielsmh) [inline src/console_cmds.cpp:1902]: Considering adding a way to summon the framerate window from console, so you can also view it on the title screen.

**Comment by ncdulo** [comment]
I really like this! I'm always interested in seeing the behind the scenes stats, especially when "breaking" the game with lots of grfs. What would the best way to test this feature out if I clone the repository and build openttd myself? I would be interested in testing this out for Linux.

**Comment by nielsmh** [comment]
@ncdulo If you can build with both the SDL and the Allegro backends, and test with both, that would be ideal. The two main things that need testing is:
1) The time graphs record with sub-millisecond precision. (Find one that displays the vertical scale at 1-2-3-4 ms or 4-8-12-16 ms, and confirm that recorded values occasionally fall between the subdivision lines.)
2) Values get recorded for the Video output element. Each frame should take longer if running at larger resolution/window size.

Additionally, confirm whether the game can run at the intended 33.33 fps with both backends. If it can't, that's a separate issue.

**Comment by nielsmh** [comment]
Tested working on Linux SDL (in a virtual machine). Interestingly, the game loop runs at about 34.1 fps here, so faster than intended.

**Comment by nielsmh** [comment]
Also seems to work with Linux Allegro. Presumably the only blocker for DOS Allegro then is whether any working DOS build environment has std::high_resolution_clock available.

**Review by TrueBrain — APPROVED** [review]
My compliments how you kept on this, changing it four times, and still going, including doing most of the testing yourself. Well done!
